### ISSUE TITLE
PyTorch analyze_videos becomes extremely slow after brief GPU usage; CPU single core pegged; occasional queue.Full & hang

### ISSUE THREAD

**Issue by anesthesiologistxia** [DeepLabCut/DeepLabCut#3123]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating System

Ubuntu 24.04(x86_64)

### DeepLabCut version

3.0.0rc9 (PyTorch engine)

### What engine are you using?

pytorch

### DeepLabCut mode

multi animal

### Device type

NVIDIA GeForce RTX 4080 SUPER (1 GPU)

### Bug description 🐛

What I’m trying to do

Run pose inference (no auto-tracking) on standard 640×480, 30 fps multi-animal mouse videos 

videos_to_analyze = [
    "/home/.../MS_Videos/animal_30fps/X417-20 CONTROL-V NOR-AB_30fps.mp4"  # 21,531 frames
]

deeplabcut.analyze_videos(
    config_path,
    videos_to_analyze,
    videotype='.mp4',
    auto_track=False,
    engine=deeplabcut.Engine.PYTORCH,
    device="cuda:0",
    # sometimes also tried: batchsize=8 or torch_kwargs={"batch_size": 8}
)

The project trains/evaluates fine (HRNet-W32 + PAF head; training and evaluate_network work and use the GPU as expected). The issue only shows up in video analysis.

Expected behavior

Stable GPU utilization during inference (batching on GPU).

Throughput consistent with model & resolution (e.g., a few hundred FPS on 640×480 with a 4080-class GPU).

CPU should not be the bottleneck / single core saturated.

Actual behavior

When analyze_videos starts, GPU utilization is high for a short time, then quickly drops to ~0–5% and stays there for the remainder.

Progress becomes very slow (e.g., ~648/21531 frames in ~3 minutes, ETA many hours), even though the model & resolution should be fast on this GPU.

CPU usage collapses to a single core pegged near 100%, suggesting a preprocessing or queueing bottleneck.

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by moutayam** [comment]
Hey @anesthesiologistxia !
Your setup might be perfectly fine, but the single CPU core 100% utilization made me wonder if PyTorch CUDA installation is not set up right. Could you please run this quick check to rule it out?
```python
import torch
print(f"CUDA available: {torch.cuda.is_available()}")
print(f"Current device: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'CPU'}")
print(f"Torch version: {torch.__version__}")
print(f"CUDA version (PyTorch): {torch.version.cuda}")
```
Also, while `analyze_videos` is running, could you check GPU memory usage with:
```bash
nvidia-smi
```
If the GPU shows 0 MB allocated during inference, that would confirm it's running on CPU despite the `device="cuda:0"` setting.
You're probably doing everything correctly, but this would help narrow down whether it's a data pipeline bottleneck or an installation problem.

Also if there is no specific need to use DeepLabCut 3.0.0rc9, I recommend creating a conda env yourself and installing PyTorch there, following the [official documentation](https://pytorch.org/get-started/locally/) for CUDA 12.9 and your OS "Windows or Linux":
```shell
conda create -n test_dlc python=3.10
conda activate test_dlc
pip install torch torchvision --index-url https://download.pytorch.org/whl/cu129 # It works on Windows and Linux
pip install deeplabcut[gui,modelzoo,wandb]==3.0.0rc10
```

I'm looking forward to hearing back from you!

Best,
Moutayam

**Comment by anesthesiologistxia** [comment]
i have try the method you mentioned above and rebuild an enviroment:

`conda create -n test_dlc python=3.10
conda activate test_dlc
pip install torch torchvision --index-url https://download.pytorch.org/whl/cu129 # It works on Windows and Linux
pip install deeplabcut[gui,modelzoo,wandb]==3.0.0rc10`
`
and when i use the code here , the output is below :
`import torch
print(f"CUDA available: {torch.cuda.is_available()}")
print(f"Current device: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'CPU'}")
print(f"Torch version: {torch.__version__}")
print(f"CUDA version (PyTorch): {torch.version.cuda}")`
**Output:**
`CUDA available: True
Current device: NVIDIA GeForce RTX 4080 SUPER
Torch version: 2.8.0+cu129
CUDA version (PyTorch): 12.9`

when i use `nvidia-smi` in my conda enviroment the output is here :
(dlc) anesthesiologist-xia@ubuntu-fo-xhh:~/Database/DeepLabCut_Data$ nvidia-smi
Sun Oct 12 09:50:26 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.107.02             Driver Version: 550.107.02     CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4080 ...    Off |   00000000:01:00.0  On |                  N/A |
|  0%   46C    P2             57W /  320W |    1974MiB /  16376MiB |      3%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A      2622      G   /usr/lib/xorg/Xorg                            207MiB |
|    0   N/A  N/A      2913      G   /usr/bin/gnome-shell                          102MiB |
|    0   N/A  N/A      3454      G   /usr/libexec/xdg-desktop-portal-gnome         113MiB |
|    0   N/A  N/A   1670146      G   /usr/bin/nautilus                              32MiB |
|    0   N/A  N/A   1762579      G   ...ste/Snipaste-2.10.2-x86_64.AppImage         13MiB |
|    0   N/A  N/A   1811795      G   ...seed-version=20251010-010036.452000         61MiB |
|    0   N/A  N/A   1823658      G   ...erProcess --variations-seed-version         83MiB |
|    0   N/A  N/A   1830861      G   /usr/share/eusoft-eudic/eudic                  54MiB |
|    0   N/A  N/A   1831422      C   ...anaconda_env_ex/envs/dlc/bin/python       1118MiB |
|    0   N/A  N/A   2670450      G   /usr/lib/xorg/Xorg                             22MiB |
+-----------------------------------------------------------------------------------------+

**but the trouble still happened, when i use the code like this:**
`videos_to_analyze = ['/home/anesthesiologist-xia/Database/DeepLabCut_Data/MS_Videos/animal_30fps/X417-20 CONTROL-V NOR-AB_30fps.mp4']
deeplabcut.analyze_videos(config_path, videos_to_analyze, videotype='.mp4',device="cuda:0",auto_track=False)`
**Output:**
Successfully loaded snapshot: /home/anesthesiologist-xia/Database/DeepLabCut_Data/Project/X401_8_X501_4_NOR-XHH-2025-10-10/dlc-models-pytorch/iteration-0/X401_8_X501_4_NOROct10-trainset95shuffle1/train/snapshot-best-010.pt
Analyzing videos with /home/anesthesiologist-xia/Database/DeepLabCut_Data/Project/X401_8_X501_4_NOR-XHH-2025-10-10/dlc-models-pytorch/iteration-0/X401_8_X501_4_NOROct10-trainset95shuffle1/train/snapshot-best-010.pt
Using scorer: DLC_HrnetW32_X401_8_X501_4_NOROct10shuffle1_snapshot_best-10
Starting to analyze /home/anesthesiologist-xia/Database/DeepLabCut_Data/MS_Videos/animal_30fps/X417-20 CONTROL-V NOR-AB_30fps.mp4
Video metadata: 
  Overall # of frames:    21531
  Duration of video [s]:  717.70
  fps:                    30.0
  resolution:             w=640, h=480

Running pose prediction with batch size 4
  4%|▍         | 868/21531 [05:11<9:17:42,  1.62s/it]
  
all i can see is that, the speed of analyze reduce from 100s/it to 1s/it, also the GPU usage from high to 0 

<img width="1218" height="249" alt="Image" src="https://github.com/user-attachments/assets/1d03a99c-113d-4c51-b600-04267cddc8b4" />

another tips are that, i can use GPU with a fast speed in using training and inference function like this :
` deeplabcut.video_inference_superanimal(
                                        videos=['/home/anesthesiologist-xia/Database/DeepLabCut_Data/MS_Videos/animal_30fps/X417-20 CONTROL-V NOR-AB_30fps.mp4'
                                                ],
                                        superanimal_name=superanimal_name,
                                        model_name="hrnet_w32",
                                        detector_name="fasterrcnn_resnet50_fpn_v2",
                                        customized_pose_checkpoint=model_name,
                                        customized_detector_checkpoint=detector_name,
                                        video_adapt = False,
                                        device="cuda:0",
                                        max_individuals=4,
                                        batch_size=4
                                       )
`

**Comment by anesthesiologistxia** [comment]
another thing i should supply is that my training model is transfer leraning from the superanimal model by using three bodyparts with 1400 frames and 10 epochs， it is may act as the posible reason？

**Comment by AlexEMG** [comment]
Indeed, I would also say that the speed is slow that your GPU is not used.

**Comment by deruyter92** [comment]
Closing this issue because of inactivity. In 3.0.0rc13 there are several improvements to the inference pipeline. If you are still encountering this problem, feel free to re-open this issue.
